libkbfs/bserver_tlf_journal.go

```diff
@@ -0,0 +1,706 @@
+}
+
+func (s *bserverTlfJournal) blockDataPath(id BlockID) string {
+	return filepath.Join(s.blockPath(id), "data")
+}
+
+func (s *bserverTlfJournal) keyServerHalfPath(id BlockID) string {
+	return filepath.Join(s.blockPath(id), "key_server_half")
+}
+
+// The functions below are for getting and setting the earliest and
+// latest ordinals.
+
+func (s *bserverTlfJournal) readOrdinalLocked(path string) (
+	journalOrdinal, error) {
+	buf, err := ioutil.ReadFile(path)
+	if err != nil {
+		return 0, err
+	}
+	return makeJournalOrdinal(string(buf))
+}
+
+func (s *bserverTlfJournal) writeOrdinalLocked(
+	path string, o journalOrdinal) error {
+	return ioutil.WriteFile(path, []byte(o.String()), 0600)
+}
+
+func (s *bserverTlfJournal) readEarliestOrdinalLocked() (
+	journalOrdinal, error) {
+	return s.readOrdinalLocked(s.earliestPath())
+}
+
+func (s *bserverTlfJournal) writeEarliestOrdinalLocked(o journalOrdinal) error {
+	return s.writeOrdinalLocked(s.earliestPath(), o)
+}
+
+func (s *bserverTlfJournal) readLatestOrdinalLocked() (journalOrdinal, error) {
+	return s.readOrdinalLocked(s.latestPath())
+}
+
+func (s *bserverTlfJournal) writeLatestOrdinalLocked(o journalOrdinal) error {
+	return s.writeOrdinalLocked(s.latestPath(), o)
+}
+
+const (
+	blockPutOp    = "blockPut"
+	addRefOp      = "addReference"
+	removeRefsOp  = "removeReferences"
+	archiveRefsOp = "archiveReferences"
+)
+
+// A bserverJournalEntry is just the name of the operation and the
+// associated block ID and contexts. Fields are exported only for
+// serialization.
+type bserverJournalEntry struct {
+	// Must be one of the four ops above.
+	Op string
+	ID BlockID
+	// Must have exactly one entry for blockPutOp and addRefOp,
+	// and at least one for removeRefsOp and archiveRefsOp.
+	Contexts []BlockContext
+}
+
+// The functions below are for reading and writing journal entries.
+
+func (s *bserverTlfJournal) readJournalEntryLocked(o journalOrdinal) (
+	bserverJournalEntry, error) {
+	p := s.journalEntryPath(o)
+	buf, err := ioutil.ReadFile(p)
+	if err != nil {
+		return bserverJournalEntry{}, err
+	}
+
+	var e bserverJournalEntry
+	err = s.codec.Decode(buf, &e)
+	if err != nil {
+		return bserverJournalEntry{}, err
+	}
+
+	return e, nil
+}
+
+// readJournalLocked reads the journal and returns a map of all the
+// block references in the journal.
+func (s *bserverTlfJournal) readJournalLocked() (
+	map[BlockID]blockRefMap, error) {
+	refs := make(map[BlockID]blockRefMap)
+
+	first, err := s.readEarliestOrdinalLocked()
+	if os.IsNotExist(err) {
+		return refs, nil
+	} else if err != nil {
+		return nil, err
+	}
+	last, err := s.readLatestOrdinalLocked()
+	if err != nil {
+		return nil, err
+	}
+
+	for i := first; i <= last; i++ {
+		e, err := s.readJournalEntryLocked(i)
+		if err != nil {
+			return nil, err
+		}
+
+		blockRefs := refs[e.ID]
+		if blockRefs == nil {
+			blockRefs = make(blockRefMap)
+			refs[e.ID] = blockRefs
+		}
+
+		switch e.Op {
+		case "put", "addReference":
```

Comment:
Why aren't these cases using the `Op` constants above?  And why is it "put" instead of "blockPut"?